presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayContainsToInnerJoin.java

```diff
@@ -137,16 +119,21 @@ public Result apply(FilterNode node, Captures captures, Context context)
         List<VariableReferenceExpression> leftInput = joinNode.getLeft().getOutputVariables();
         List<VariableReferenceExpression> rightInput = joinNode.getRight().getOutputVariables();
         RowExpression filterExpression = node.getPredicate();
-        RowExpression arrayContainsExpression = getCandidateArrayContainsExpression(filterExpression, leftInput, rightInput);
+        FunctionResolution functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+
+        RowExpression arrayContainsExpression = getCandidateArrayContainsExpression(functionResolution, filterExpression, leftInput, rightInput);
         if (arrayContainsExpression == null) {
             return Result.empty();
         }
         List<RowExpression> andConjuncts = extractConjuncts(filterExpression);
         List<RowExpression> remainingConjuncts = andConjuncts.stream().filter(x -> !x.equals(arrayContainsExpression)).collect(toImmutableList());
 
-        VariableReferenceExpression array = (VariableReferenceExpression) ((CallExpression) arrayContainsExpression).getArguments().get(0);
-        VariableReferenceExpression element = (VariableReferenceExpression) ((CallExpression) arrayContainsExpression).getArguments().get(1);
-        boolean arrayAtLeftInput = leftInput.contains(array);
+        RowExpression array = ((CallExpression) arrayContainsExpression).getArguments().get(0);
+        RowExpression element = ((CallExpression) arrayContainsExpression).getArguments().get(1);
+
+        checkState(element instanceof VariableReferenceExpression, "Argument to CONTAINS is not a column");
```

Comment:
The error message is misleading, we can have argument of contains to be expression, just we do not allow here. Maybe we can leave this alone as this is only a safety check.

Reply:
Maybe also add assert for `array`?

Reply:
added